### ISSUE TITLE
Motion + Raspberry Pi NoIR camera is underexposed

### ISSUE THREAD

**Issue by alligitor** [Motion-Project/motion#582]
1. version [x.y.z, hash, other]: Version 4.0, Copyright 2000-2016
2. installed as a package or compiled from sources [deb, rpm, git, other]: package on arm
3. standalone or part of third party [motion, MotionEyeOS, other]: stand alone
4. video stream source [V4L (card or USB), net cam (mjpeg, rtsp, other), mmal]: V4L 
5. hardware [x86, ARM, other]: ARM
6. operating system [Linux (which), FreeBSD, other]: LINUX

I have a bunch of raspberry pis and NoIR cameras to stream video to my browser.  the video is underexposed and much darker than what raspivid provides, or other network cameras such as Foscam provide.

I am no 100% sure but I think this is more of a code problem with motion not setting up the camera parameters correctly than a camera or configuration problem, hence my post here.  

Is this a problem with motion not setting up exposure parameters on the NoIR camera properly?
Is this a problem with the RP Camera's firmware not working?

here are my settings:
 ```
<– back

Camera 0 -- RP-Camera

    daemon = on
    process_id_file = /var/run/motion/motion.pid
    setup_mode = off
    camera_name = RP-Camera
    logfile = (not defined)
    log_level = 6
    log_type = all
    videodevice = /dev/video0
    v4l2_palette = 17
    input = -1
    norm = 0
    frequency = 0
    rotate = 0
    width = 1024
    height = 768
    framerate = 100
    minimum_frame_time = 0
    netcam_url = (not defined)
    netcam_userpass = (not defined)
    netcam_keepalive = off
    netcam_proxy = (not defined)
    netcam_tolerant_check = off
    rtsp_uses_tcp = on
    auto_brightness = off
    brightness = 0
    contrast = 0
    saturation = 0
    hue = 0
    power_line_frequency = -1
    roundrobin_frames = 1
    roundrobin_skip = 1
    switchfilter = off
    threshold = 1500
    threshold_tune = off
    noise_level = 32
    noise_tune = on
    despeckle_filter = EedDl
    area_detect = (not defined)
    mask_file = (not defined)
    smart_mask_speed = 0
    lightswitch = 0
    minimum_motion_frames = 1
    pre_capture = 0
    post_capture = 0
    event_gap = 60
    max_movie_time = 0
    emulate_motion = off
    output_pictures = off
    output_debug_pictures = off
    quality = 75
    camera_id = 0
    picture_type = jpeg
    ffmpeg_output_movies = off
    ffmpeg_output_debug_movies = off
    ffmpeg_timelapse = 0
    ffmpeg_timelapse_mode = daily
    ffmpeg_bps = 500000
    ffmpeg_variable_bitrate = 0
    ffmpeg_video_codec = mpeg4
    ffmpeg_duplicate_frames = off
    use_extpipe = off
    extpipe = (not defined)
    snapshot_interval = 0
    locate_motion_mode = off
    locate_motion_style = box
    text_right = %Y-%m-%d\n%T-%q
    text_left = (not defined)
    text_changes = off
    text_event = %Y%m%d%H%M%S
    text_double = off
    exif_text = (not defined)
    target_dir = /tmp
    snapshot_filename = %v-%Y%m%d%H%M%S-snapshot
    picture_filename = %v-%Y%m%d%H%M%S-%q
    movie_filename = %v-%Y%m%d%H%M%S
    timelapse_filename = %Y%m%d-timelapse
    ipv6_enabled = off
    stream_port = 8081
    stream_quality = 50
    stream_motion = off
    stream_maxrate = 100
    stream_localhost = off
    stream_limit = 0
    stream_auth_method = 2
    stream_authentication = visitor:visitor
    stream_preview_scale = 25
    stream_preview_newline = off
    webcontrol_port = 8080
    webcontrol_localhost = off
    webcontrol_html_output = on
    webcontrol_authentication = (not defined)
    track_type = 0
    track_auto = off
    track_port = (not defined)
    track_motorx = 0
    track_motorx_reverse = off
    track_motory = 0
    track_motory_reverse = off
    track_maxx = 0
    track_minx = 0
    track_maxy = 0
    track_miny = 0
    track_homex = 128
    track_homey = 128
    track_iomojo_id = 0
    track_step_angle_x = 10
    track_step_angle_y = 10
    track_move_wait = 10
    track_speed = 255
    track_stepsize = 40
    quiet = on
    on_event_start = (not defined)
    on_event_end = (not defined)
    on_picture_save = (not defined)
    on_motion_detected = (not defined)
    on_area_detected = (not defined)
    on_movie_start = (not defined)
    on_movie_end = (not defined)
    on_camera_lost = (not defined)
    sql_log_picture = on
    sql_log_snapshot = on
    sql_log_movie = off
    sql_log_timelapse = off
    sql_query = sql_query insert into security(camera, filename, frame, file_type, time_stamp, event_time_stamp) values('%t', '%f', '%q', '%n', '%Y-%m-%d %T', '%C')
    database_type = (not defined)
    database_dbname = (not defined)
    database_host = localhost
    database_user = (not defined)
    database_password = (not defined)
    database_port = 0
    database_busy_timeout = 0
    video_pipe = (not defined)
    motion_video_pipe = (not defined)
    camera = No threads
    thread = No threads
    camera_dir = /etc/conf.d

<– back 
```

**Comment by Mr-Dave** [comment]
You may want to consider using some of the Motion configuration options.  The configuration posted has most options disabled.

You may also have to use the v4l2-ctl utility to set pi camera options prior to starting Motion.  

Finally, there is also the option to the MMAL features but only if you upgrade.  The MMAL options can not be part of the packages provided by apt.  

Please consult the [support](https://motion-project.github.io/motion_support.html) page, [Documentation](https://motion-project.github.io/motion_config.html) pages and if upgrading the [Downloads](https://motion-project.github.io/motion_download.html) page.

**Comment by alligitor** [comment]
I  played with some configuration options such as brightness.  it only brightens the underexposed image so it looses detail
setting v4l2 parameters didn't work for two reasons:  
1- I had to restart motion for them to take effect
2- even after a restart, the setting works until the environment lighting changes. i.e night->day

I downloaded latest motion and compiled on my RPi.  this comes from my syslog:

> "Dec 18 20:48:03 RP-Camera motion: [1:ml1:RP-Camera] [NTC] [VID] v4l2_get_capability: #012------------------------#012cap.driver: "bm2835 mmal"#012cap.card: "mmal service 16.1"#012cap.bus_info: "platform:bcm2835-v4l2"#012cap.capabilities=0x85200005#012-----"

I still think this is a but or feature request for motion.  I believe, this is how it should work:
motion evaluates an image from the camera
evaluate its brightness
manipulate options such as shutter speed, ISO, exposure to get a better image.

I admit I am not too familiar with this, but to me it sounds like it would be motion's job to set the camera parameters using v4l2 parameters.

I'm willing to experiment with options or even code changes if somebody gives some pointers to where in the code to look

**Comment by Mr-Dave** [comment]
I am not seeing this as a bug.  Motion does set the parameters for the common options such as brightness, hue, contrast, etc.  

If you identify within the `v4l2-cntl` parameters the particular ones that provides the desired effect, then they can be contemplated for inclusion as Motion configuration parameters.  Note that the particular items that you referenced (shutter speed and ISO) are not options that even _CAN_ be manipulated via a v4l2 device.  Exposure is listed as a possible option in the [reference.](https://www.linuxtv.org/downloads/legacy/video4linux/API/V4L2_API/spec/ch01s08.html) but  particular attention should be paid to the first paragraph where it indicates not all devices have all of the options and even if they have the same option, different values apply.

If you would like to test and manipulate the options, the particular section of code that sets these items is [here](https://github.com/Motion-Project/motion/blob/master/video_v4l2.c#L751).  An easy hack to see if exposure would work would be to replace something like `V4L2_CID_HUE` with `V4L2_CID_EXPOSURE` and then any Motion configuration option change for `hue' would actually be changing the exposure setting.

Finally, since you have now upgraded, a much better option is probably to use the MMAL functionality instead of the videodevice as described in the [guide](https://motion-project.github.io/motion_config.html#Basic_Setup).  The MMAL option allows you to pass in whatever raspivid options you need via the [mmalcam_control_params](https://motion-project.github.io/motion_config.html#Options_Raspi_Cameras).  As indicated previously, the MMAL functionality is not available when Motion is installed via apt due to how Raspbian packages the software.

**Comment by alligitor** [comment]
also, I agree that what I requested is not a bug.  As a feature request however, has some value.
Perhaps instead of spending cpu cycles adjusting the brightness within motion software, motion could evaluate a frame and adjust the exposure setting to let the camera FW handle computation.
----------------------
I tested two values in v4l2_picture_controls by trying each of the two lines separately:
    v4l2_set_control(vid_source, V4L2_CID_EXPOSURE, 0);
    v4l2_set_control(vid_source, V4L2_CID_EXPOSURE, 24);
and the picture was darker (0)/brighter (24).  So it works.  The two values come from rasp NoIR camera.

Based on my quick read of the code, to formally add this option, what would remain is adding an "exposure" member to video_dev and config structures and add a new text string to config file processing.

Also, v4l API has options for some other interesting parameters such as metering mode, whitebalance, etc. that are probably easy to add and give a lot of flexibility.

i'll play more with mmal functions and update the thread.
------------------------
